csrc/kernel_cache.cpp

```diff
@@ -779,10 +779,31 @@ FusionKernelRuntime* FusionExecutorCache::getKernelRuntimeFor(
   // that whenever we encounter a new set of input shapes we segment and compile
   // a new FusionKernelRuntime.
   if (!isOptionDisabled(DisableOption::KernelReuse)) {
+    std::optional<bool> should_not_be_segmented = std::nullopt;
     auto reuse_it = std::find_if(
         kernel_runtimes.begin(),
         kernel_runtimes.end(),
-        [&args, &new_heuristics, &forced_index_type](auto& kernel_runtime) {
+        [&should_not_be_segmented, &args, &new_heuristics, &forced_index_type](
```

Comment:
can you comment about the general logic here? The logic in my mind is (1) search for unsegmented runtime cache, (2) if doesn't exist try to schedule as unsegmented, (3) if can't, search for segmented runtime.

Reply:
Added some comments to the code. This block is getting the first `FusionKernelRuntime` with the same concretization and for which we use the same heuristics in each segment. That is what happens, but it happens all at once here:
1. If `should_not_be_segmented` is not yet set, initialize it by re-using the concretized fusion in `kernel_runtime`. This is done to prevent needing to perform a redundant concretization before entering this loop.
2. If `should_not_be_segmented` is true, then filter out segmented `FusionKernelRuntime`s (this implements your (1) and (2).
3. If `should_not_be_segmented` is false, include this in the search. It will be returned as `*reuse_it` if all segment heuristics match.